timemachine/potentials/potentials.py

```diff
@@ -0,0 +1,270 @@
+
+@dataclass
+class NonbondedPairListNegated(Potential):
+    idxs: NDArray[np.int32]
+    rescale_mask: NDArray[np.float64]
+    beta: float
+    cutoff: float
+
+    def __call__(self, conf: Conf, params: Params, box: Optional[Box]) -> float | Array:
+        vdW, electrostatics = nonbonded.nonbonded_on_specific_pairs(
+            conf, params, box, self.idxs, self.beta, self.cutoff, self.rescale_mask
+        )
+        U = jnp.sum(vdW) + jnp.sum(electrostatics)
+        return -U
+
+    @classmethod
+    def _custom_ops_class_name(cls, precision):
+        return f"NonbondedPairList_{get_custom_ops_class_name_suffix(precision)}_negated"
+
+
+@dataclass
+class NonbondedPairListPrecomputed(Potential):
+    """
+    This implements a pairlist with precomputed parameters. It differs from the regular NonbondedPairlist in that it
+    expects params of the form s0*q_ij, s_ij, s1*e_ij, and w_offsets_ij, where s are the scaling factors and combining
+    rules have already been applied.
+
+    Note that you should not use this class to implement exclusions (that are later cancelled out by AllPairs) since the
+    floating point operations are different in python vs C++.
+    """
+
+    idxs: NDArray[np.int32]
+    beta: float
+    cutoff: float
+
+    def __call__(self, conf: Conf, params: Params, box: Optional[Box]) -> float | Array:
+        vdW, electrostatics = nonbonded.nonbonded_on_precomputed_pairs(
+            conf, params, box, self.idxs, self.beta, self.cutoff
+        )
+        return jnp.sum(vdW) + jnp.sum(electrostatics)
+
+
+@dataclass
+class SummedPotential(Potential):
+    potentials: Sequence[Potential]
+    params_init: Sequence[NDArray]
+
+    def __post_init__(self):
+        if len(self.potentials) != len(self.params_init):
+            raise ValueError("number of potentials != number of parameter arrays")
+
+    def __call__(self, conf: Conf, params: Params, box: Optional[Box]) -> float | Array:
+        shapes = [ps.shape for ps in self.params_init]
+        return summed.summed_potential(conf, params, box, self.potentials, shapes)
+
+    def to_gpu(self, precision: Precision) -> "SummedPotentialGpuImplWrapper":
+        impls = [p.to_gpu(precision).unbound_impl for p in self.potentials]
+        sizes = [ps.size for ps in self.params_init]
+        return SummedPotentialGpuImplWrapper(custom_ops.SummedPotential(impls, sizes))
+
+
+@dataclass
+class SummedPotentialGpuImplWrapper(GpuImplWrapper):
+    """Handles flattening parameters before passing to kernel to provide a nicer interface"""
+
+    def call_with_params_list(self, conf: Conf, params: Sequence[Params], box: Box) -> float:
+        params_flat = jnp.concatenate([ps.reshape(-1) for ps in params])
+        res = jax_interface.call_unbound_impl(self.unbound_impl, conf, params_flat, box)
+        return cast(float, res)
+
+    def bind_params_list(self, params: Params) -> "BoundGpuImplWrapper":
+        params_flat = np.concatenate([ps.reshape(-1) for ps in params])
```

Comment:
np.concatenate or jnp.concatenate?

Reply:
Good catch. Using `np` instead of `jnp` here was actually intentional, to emphasize that it doesn't make sense to differentiate `bind_params_list` with respect to `params`.